packages/hash/frontend/src/blocks/page/BlockContextMenu/BlockContextMenu.tsx

```diff
@@ -0,0 +1,223 @@
+import React, { useRef, forwardRef, useMemo } from "react";
+
+import { useKey } from "rooks";
+import { EntityStore, isBlockEntity } from "@hashintel/hash-shared/entityStore";
+
+import { EditorView } from "prosemirror-view";
+import { Schema } from "prosemirror-model";
+import { Box, Divider, Menu, Typography } from "@mui/material";
+import { bindMenu } from "material-ui-popup-state";
+import { PopupState } from "material-ui-popup-state/hooks";
+import { format } from "date-fns";
+import {
+  faAdd,
+  faArrowRight,
+  faLink,
+  faRefresh,
+} from "@fortawesome/free-solid-svg-icons";
+import {
+  faCopy,
+  faMessage,
+  faTrashCan,
+} from "@fortawesome/free-regular-svg-icons";
+import { getBlockDomId } from "../BlockView";
+import { BlockSuggesterProps } from "../createSuggester/BlockSuggester";
+
+import { BlockLoaderInput } from "./BlockLoaderInput";
+import { useUsers } from "../../../components/hooks/useUsers";
+import { FontAwesomeIcon } from "../../../shared/icons";
+import { BlockContextMenuItem } from "./BlockContextMenuItem";
+import { LoadEntityMenuContent } from "./LoadEntityMenuContent";
+import { BlockListMenuContent } from "./BlockListMenuContent";
+
+type BlockContextMenuProps = {
+  popupState: PopupState;
+  blockSuggesterProps: BlockSuggesterProps;
+  entityId: string | null;
+  entityStore: EntityStore;
+  view: EditorView<Schema>;
+};
+
+const LOAD_BLOCK_ENTITY_UI = "hash-load-entity-ui";
+
+export const BlockContextMenu = forwardRef<
+  HTMLDivElement,
+  BlockContextMenuProps
+>(({ popupState, blockSuggesterProps, entityId, entityStore }, ref) => {
+  const blockData = entityId ? entityStore.saved[entityId] : null;
+  const { data: users } = useUsers();
+  const setEntityMenuItemRef = useRef<HTMLLIElement>(null);
+  const swapBlocksMenuItemRef = useRef<HTMLLIElement>(null);
+
+  if (blockData && !isBlockEntity(blockData)) {
+    throw new Error("BlockContextMenu linked to non-block entity");
+  }
+
+  const menuItems = useMemo(() => {
+    const hasChildEntity =
+      Object.keys(blockData?.properties.entity?.properties ?? {}).length > 0;
+    const items = [
```

Comment:
The array is there to reduce code duplication. I think it's easier to add/remove/update items in the menu using this approach. The array wasn't introduced in this PR though. It was just outside the component, I moved it into the component because I needed to access props.